DigitalLearningSolutions.Web/Controllers/Register/ClaimAccountController.cs

```diff
@@ -70,14 +94,81 @@ public IActionResult CompleteRegistration()
             claimAccountService.ConvertTemporaryUserToConfirmedUser(
                 model.UserId,
                 model.CentreId,
-                model.CentreSpecificEmail
+                model.Email
             );
 
             TempData.Clear();
+
             return View("Confirmation", model);
         }
 
-        private ClaimAccountViewModel? GetViewModelIfValidParameters(string email, string code)
+        [Authorize(Policy = CustomPolicies.BasicUser)]
+        [HttpGet]
+        public IActionResult LinkDlsAccount(string email, string code)
+        {
+            var loggedInUserId = User.GetUserIdKnownNotNull();
+            var model = GetViewModelIfValidParameters(email, code, loggedInUserId);
+            var actionResult = ValidateClaimAccountViewModelForLinkingAccounts(loggedInUserId, model);
+
+            TempData.Set(model);
+
+            if (actionResult != null)
+            {
+                return actionResult;
+            }
+
+            return View(model);
+        }
+
+        [Authorize(Policy = CustomPolicies.BasicUser)]
+        [HttpPost]
+        public IActionResult LinkDlsAccount()
+        {
+            var model = TempData.Peek<ClaimAccountViewModel>()!;
+            var loggedInUserId = User.GetUserIdKnownNotNull();
+            var actionResult = ValidateClaimAccountViewModelForLinkingAccounts(loggedInUserId, model);
+
+            if (actionResult != null)
+            {
+                return actionResult;
+            }
+
+            claimAccountService.LinkAccount(
+                model.UserId,
+                loggedInUserId,
+                model.CentreId
+            );
+
+            TempData.Clear();
+
+            return View("AccountsLinked", model);
+        }
+
+        [HttpGet]
+        public IActionResult WrongUser()
+        {
+            var model = TempData.Peek<ClaimAccountViewModel>();
+
+            TempData.Clear();
+
+            return model == null ? NotFound() as IActionResult : View(model);

```

Comment:
Better to pass the model into this rather than using temp data - currently I think refreshing the page will give you a NotFound.

Reply:
With no TempData comes no ~responsibility~ NotFound page